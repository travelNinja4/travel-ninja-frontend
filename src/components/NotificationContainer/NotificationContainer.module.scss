/**
------------------------------------------------------
styles specific to the NotificationContainer component
------------------------------------------------------
**/

/* Wrapper for positions */
.notificationWrapper {
  position: fixed;
  z-index: 9999;
  display: flex;
  flex-direction: column;
  gap: 12px;
  pointer-events: none; // allow clicks to pass through empty space
}

/* Notification box */
.notification {
  pointer-events: auto; // clickable
  display: flex;
  align-items: flex-start;
  gap: 12px;
  min-width: 280px;
  max-width: 360px;
  padding: 14px 16px;
  border-radius: 8px;
  color: var(--color-white);
  font-size: 14px;
  box-shadow: 0 6px 18px rgba(0, 0, 0, 0.15);
  position: relative;
  animation: slideIn 0.3s ease forwards;
  border-left: 4px solid;
}

/* Icon */
.notificationIcon {
  display: flex;
  align-items: center;
  justify-content: center;
  margin-top: 3%;
}

/* Text block */
.notificationText {
  flex: 1;
}

.notificationTitle {
  font-weight: 600;
  font-size: 1rem;
  margin-bottom: 2px;
  display: block;

  &.successTitle {
    color: var(--notification-success-title);
  }

  &.errorTitle {
    color: var(--notification-error-title);
  }

  &.warningTitle {
    color: var(--notification-warning-title);
  }

  &.infoTitle {
    color: var(--notification-info-title);
  }
}

.notificationMessage {
  margin: 0;
  font-size: 0.875rem;
  line-height: 1.4;

  &.successNotificationMessage {
    color: var(--notification-success-message);
  }

  &.errorNotificationMessage {
    color: var(--notification-error-message);
  }

  &.warningNotificationMessage {
    color: var(--notification-warning-message);
  }

  &.infoNotificationMessage {
    color: var(--notification-info-message);
  }
}

/* Close button */
.notificationClose {
  cursor: pointer;
  margin-left: 8px;
}

/* Type colors */
.success {
  border-left-color: var(--notification-success-border);
  background: linear-gradient(
    135deg,
    var(--notification-success-bg-start) 0%,
    var(--notification-success-bg-end) 100%
  );
}

.error {
  border-left-color: var(--notification-error-border);
  background: linear-gradient(
    135deg,
    var(--notification-error-bg-start) 0%,
    var(--notification-error-bg-end) 100%
  );
}

.warning {
  border-left-color: var(--notification-warning-border);
  background: linear-gradient(
    135deg,
    var(--notification-warning-bg-start) 0%,
    var(--notification-warning-bg-end) 100%
  );
}

.info {
  border-left-color: var(--notification-info-border);
  background: linear-gradient(
    135deg,
    var(--notification-info-bg-start) 0%,
    var(--notification-info-bg-end) 100%
  );
}

/* Progress bar container */
.notificationProgress {
  position: absolute;
  bottom: 0;
  left: 0;
  width: 100%;
  height: 4px;
  background: var(--notification-progress-bg);
  border-bottom-left-radius: 8px;
  border-bottom-right-radius: 8px;
  overflow: hidden;
}

/* Animated progress bar */
.notificationProgressBar {
  height: 100%;
  background: var(--notification-progress-bar);
  width: 100%;
  animation-name: progressBar;
  animation-timing-function: linear;
  animation-fill-mode: forwards;
}

/* Position wrappers */
.top-right {
  top: 20px;
  right: 20px;
  align-items: flex-end;
}

.top-left {
  top: 20px;
  left: 20px;
  align-items: flex-start;
}

.bottom-right {
  bottom: 20px;
  right: 20px;
  align-items: flex-end;
}

.bottom-left {
  bottom: 20px;
  left: 20px;
  align-items: flex-start;
}

.top-center {
  top: 20px;
  left: 50%;
  transform: translateX(-50%);
  align-items: center;
}

/* Animations */
@keyframes progressBar {
  from {
    width: 100%;
  }
  to {
    width: 0%;
  }
}

@keyframes slideIn {
  from {
    opacity: 0;
    transform: translateY(-10px);
  }
  to {
    opacity: 1;
    transform: translateY(0);
  }
}
