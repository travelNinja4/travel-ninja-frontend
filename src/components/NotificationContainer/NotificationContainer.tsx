/**
 * NotificationContainer renders active notifications on screen with their styles, positions, progress bars, and close buttons, while handling display and dismissal animations.
 *
 * @example
 * ```tsx
 * import NotificationContainer from '@src/components/NotificationContainer'
 *
 * export default function NotificationContainer() {
 * const [notifications, setNotifications] = useState<Notification[]>([
 *     {
 *       id: 1,
 *       type: 'success',
 *       title: 'Success!',
 *       message: 'Your action was completed successfully.',
 *       duration: 3000,
 *       showProgress: true,
 *       position: 'top-right',
 *     },
 *   ]);
 *
 *   const removeNotification = (id: number) => {
 *     setNotifications((prev) => prev.filter((n) => n.id !== id));
 *   };
 *   return (
 *     <NotificationContainer
 *       notifications={notifications}
 *       removeNotification={removeNotification}
 *     />
 * }
 * ```
 */
'use client';

import { useEffect, useState } from 'react';
import { CircleCheck, CircleX, Info, TriangleAlert, X } from 'lucide-react';
import clsx from 'clsx';
import Typography from '@/components/Typography';
import { NOTIFICATION_TYPES } from '@/constants/strings';
import styles from './NotificationContainer.module.scss';

/**
 * Define the props available for the NotificationContainer component.
 */
interface NotificationContainerProps {
  /**
   * The list of notifications to be displayed.
   *
   * Each notification contains metadata such as `id`, `type`,
   * `title`, `message`, optional `duration`, progress bar visibility, and position.
   */
  notifications: Notification[];

  /**
   * A callback function to remove a notification.
   *
   * Called when the user closes a notification manually
   * or when the notification auto-dismisses after its `duration`.
   *
   * @param id The unique identifier of the notification to remove.
   */
  removeNotification: (id: number) => void;
}

export interface Notification {
  id: number;
  type: NotificationType;
  title: string;
  message: string;
  duration?: number;
  showProgress?: boolean;
  position?: string;
}

const iconMap = {
  success: <CircleCheck height={24} width={24} color="var(--notification-success-icon)" />,
  error: <CircleX height={24} width={24} color="var(--notification-error-icon)" />,
  warning: <TriangleAlert height={24} width={24} color="var(--notification-warning-icon)" />,
  info: <Info height={24} width={24} color="var(--notification-info-icon)" />,
};

type NotificationType = 'success' | 'error' | 'warning' | 'info';

interface ClassNames {
  title: string;
  message: string;
}

export default function NotificationContainer({
  notifications,
  removeNotification,
}: NotificationContainerProps) {
  const [activeNotifications, setActiveNotifications] = useState<Notification[]>([]);

  useEffect(() => {
    setActiveNotifications(notifications);
  }, [notifications]);

  useEffect(() => {
    const timers = notifications.map((n) => {
      if (n.duration) {
        return setTimeout(() => removeNotification(n.id), n.duration);
      }
      return null;
    });

    return () => {
      timers.forEach((t) => t && clearTimeout(t));
    };
  }, [notifications, removeNotification]);

  // Group by position
  const positions = ['top-right', 'top-left', 'bottom-right', 'bottom-left', 'top-center'];

  const getNotificationClassNames = (type: NotificationType): ClassNames => {
    switch (type) {
      case NOTIFICATION_TYPES.SUCCESS:
        return {
          title: clsx(styles.notificationTitle, styles.successTitle),
          message: clsx(styles.notificationMessage, styles.successNotificationMessage),
        };
      case NOTIFICATION_TYPES.ERROR:
        return {
          title: clsx(styles.notificationTitle, styles.errorTitle),
          message: clsx(styles.notificationMessage, styles.errorNotificationMessage),
        };
      case NOTIFICATION_TYPES.WARNING:
        return {
          title: clsx(styles.notificationTitle, styles.warningTitle),
          message: clsx(styles.notificationMessage, styles.warningNotificationMessage),
        };
      case NOTIFICATION_TYPES.INFO:
        return {
          title: clsx(styles.notificationTitle, styles.infoTitle),
          message: clsx(styles.notificationMessage, styles.infoNotificationMessage),
        };
      default:
        return {
          title: styles.notificationTitle,
          message: styles.notificationMessage,
        };
    }
  };

  return (
    <div data-testid="NotificationContainerTest">
      {positions.map((pos) => (
        <div key={pos} className={clsx(styles.notificationWrapper, styles[pos])}>
          {activeNotifications
            .filter((n) => (n.position || 'top-right') === pos)
            .map((n) => {
              const classNames = getNotificationClassNames(n.type);
              return (
                <div key={n.id} className={clsx(styles.notification, styles[n.type], styles.show)}>
                  <div className={styles.notificationIcon}>{iconMap[n.type]}</div>
                  <div className={styles.notificationText}>
                    <Typography tag="h4" className={classNames.title}>
                      {n.title}
                    </Typography>
                    <Typography tag="p" className={classNames.message}>
                      {n.message}
                    </Typography>
                  </div>

                  <div
                    className={styles.notificationClose}
                    data-testid="notification-close-btn"
                    onClick={() => removeNotification(n.id)}
                  >
                    <X height={24} width={24} color="var(--notification-close-icon)" />
                  </div>
                  {n.showProgress && n.duration && (
                    <div className={styles.notificationProgress}>
                      <div
                        className={styles.notificationProgressBar}
                        data-testid="progress-bar"
                        style={{ animationDuration: `${n.duration}ms` }}
                      />
                    </div>
                  )}
                </div>
              );
            })}
        </div>
      ))}
    </div>
  );
}
