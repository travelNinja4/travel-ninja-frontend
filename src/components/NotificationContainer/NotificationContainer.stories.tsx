import type { Meta, StoryObj } from '@storybook/nextjs-vite';
/**
import { fn } from 'storybook/test';
*/
import NotificationContainer from './NotificationContainer'

const meta = {
  title: 'components/NotificationContainer',
  component: NotificationContainer,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    label: { 
        control: 'text', 
        description: 'Sample label for the component' 
        },
  },
  args: { 
    /** onClick: fn() **/
    },
} satisfies Meta<typeof NotificationContainer>;

export default meta;
type Story = StoryObj<typeof meta>;