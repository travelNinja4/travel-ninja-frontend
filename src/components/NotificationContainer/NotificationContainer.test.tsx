import '@testing-library/jest-dom';
import { render, screen, fireEvent, cleanup } from '@testing-library/react';
import type { Notification } from './NotificationContainer';
import NotificationContainer from './NotificationContainer';

describe('Home', () => {
  /** Base props for NotificationContainer **/
  const baseNotification: Notification = {
    id: 1,
    type: 'success',
    title: 'Test Notification',
    message: 'This is a test notification',
    duration: 3000,
    showProgress: true,
    position: 'top-right',
  };

  let removeNotificationMock: jest.Mock;

  beforeEach(() => {
    removeNotificationMock = jest.fn();
    jest.useFakeTimers();
  });

  afterEach(() => {
    cleanup();
    jest.clearAllTimers();
    jest.useRealTimers();
  });

  const renderContainer = (notifications: Notification[] = [baseNotification]) => {
    return render(
      <NotificationContainer
        notifications={notifications}
        removeNotification={removeNotificationMock}
      />,
    );
  };

  /**
   * @test
   * Renders the NotificationContainer component without throwing an error
   * and checks if the NotificationContainerTest text is present in the document.
   *
   * This ensures that the component loads correctly with the base props.
   */
  it('renders without crashing', () => {
    renderContainer();
    expect(screen.getByTestId('NotificationContainerTest')).toBeInTheDocument();
  });

  /**
   * @test
   * Captures a snapshot of the NotificationContainer component with base props
   * and ensures it matches the previously stored snapshot.
   *
   * This helps detect any unexpected UI changes.
   */
  it('matches snapshot', () => {
    const { asFragment } = renderContainer();
    expect(asFragment()).toMatchSnapshot();
  });

  it('auto-dismisses notification after duration', () => {
    renderContainer();
    jest.advanceTimersByTime(3000);
    expect(removeNotificationMock).toHaveBeenCalledWith(1);
  });

  it('renders notifications of all types', () => {
    const types: Notification['type'][] = ['success', 'error', 'warning', 'info'];
    types.forEach((type, idx) => {
      renderContainer([{ ...baseNotification, id: idx, type }]);
      expect(screen.getByText('Test Notification')).toBeInTheDocument();
      cleanup(); // clear previous notification before next iteration
    });
  });

  it('renders notifications in all positions', () => {
    const positions: Notification['position'][] = [
      'top-right',
      'top-left',
      'bottom-right',
      'bottom-left',
      'top-center',
    ];
    positions.forEach((position, idx) => {
      renderContainer([{ ...baseNotification, id: idx, position }]);
      expect(screen.getByTestId('NotificationContainerTest')).toBeInTheDocument();
      cleanup();
    });
  });

  it('renders progress bar only if showProgress is true', () => {
    // With progress
    renderContainer();
    expect(screen.getByTestId('progress-bar')).toBeInTheDocument();
    cleanup();

    // Without progress
    renderContainer([{ ...baseNotification, showProgress: false }]);
    expect(screen.queryByTestId('progress-bar')).not.toBeInTheDocument();
  });

  it('calls removeNotification when close button is clicked', () => {
    renderContainer();
    const closeBtn = screen.getByTestId('notification-close-btn');
    fireEvent.click(closeBtn);
    expect(removeNotificationMock).toHaveBeenCalledWith(1);
  });
});
