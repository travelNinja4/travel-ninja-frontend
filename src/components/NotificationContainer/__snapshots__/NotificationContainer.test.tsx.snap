// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Home matches snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="NotificationContainerTest"
  >
    <div
      class="notificationWrapper top-right"
    >
      <div
        class="notification success show"
      >
        <div
          class="notificationIcon"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-circle-check"
            fill="none"
            height="24"
            stroke="var(--notification-success-icon)"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
            <path
              d="m9 12 2 2 4-4"
            />
          </svg>
        </div>
        <div
          class="notificationText"
        >
          <h4
            class="typography h4 align-left notificationTitle successTitle"
            data-testid="TypographyTest"
          >
            Test Notification
          </h4>
          <p
            class="typography p align-left notificationMessage successNotificationMessage"
            data-testid="TypographyTest"
          >
            This is a test notification
          </p>
        </div>
        <div
          class="notificationClose"
          data-testid="notification-close-btn"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-x"
            fill="none"
            height="24"
            stroke="var(--notification-close-icon)"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 6 6 18"
            />
            <path
              d="m6 6 12 12"
            />
          </svg>
        </div>
        <div
          class="notificationProgress"
        >
          <div
            class="notificationProgressBar"
            data-testid="progress-bar"
            style="animation-duration: 3000ms;"
          />
        </div>
      </div>
    </div>
    <div
      class="notificationWrapper top-left"
    />
    <div
      class="notificationWrapper bottom-right"
    />
    <div
      class="notificationWrapper bottom-left"
    />
    <div
      class="notificationWrapper top-center"
    />
  </div>
</DocumentFragment>
`;
